File: src/main/java/org/modelingvalue/nelumbo/syntax/PatternResult.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2026 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo.syntax;

import org.modelingvalue.collections.List;
import org.modelingvalue.collections.Map;
import org.modelingvalue.collections.Set;
import org.modelingvalue.collections.mutable.MutableList;
import org.modelingvalue.collections.mutable.MutableMap;
import org.modelingvalue.collections.util.Pair;
import org.modelingvalue.nelumbo.AstElement;
import org.modelingvalue.nelumbo.Node;
import org.modelingvalue.nelumbo.Type;
import org.modelingvalue.nelumbo.Variable;
import org.modelingvalue.nelumbo.patterns.Functor;
import org.modelingvalue.nelumbo.patterns.RepetitionPattern;

public final class PatternResult implements ParseExceptionHandler {

    private final MutableList<AstElement>         elements;
    private final Parser                          parser;
    private final ParseContext                    context;
    private final MutableList<Pair<Token, Token>> splitted;
    private final MutableList<Pair<Token, Token>> merged;
    private final MutableMap<Variable, Type>      typeArgs;

    private Functor                               functor;
    private ParseState                            state;
    private Integer                               leftPrecedence;
    private Set<RepetitionPattern>                endRepetitions;
    private Token                                 nextToken;
    private boolean                               hasLeft;

    public PatternResult(Parser parser, ParseContext context) {
        this.parser = parser;
        this.context = context;
        elements = MutableList.of(List.of());
        splitted = MutableList.of(List.of());
        merged = MutableList.of(List.of());
        typeArgs = MutableMap.of(Map.of());
        endRepetitions = Set.of();
    }

    public Token addSplit(Token original, Token split) {
        splitted.add(Pair.of(original, split));
        return split;
    }

    public Token addMerge(Token original, Token merge) {
        merged.add(Pair.of(original, merge));
        return merge;
    }

    public Parser parser() {
        return parser;
    }

    public ParseContext context() {
        return context;
    }

    public Functor functor() {
        return functor;
    }

    public ParseState state() {
        return state;
    }

    public Set<RepetitionPattern> endRepetitions() {
        return endRepetitions;
    }

    public Token nextToken() {
        return nextToken;
    }

    public int leftPrecedence() {
        return leftPrecedence;
    }

    public void endPostParse(Functor functor, Token nextToken, Integer leftPrecedence) {
        this.endRepetitions = Set.of();
        this.functor = functor;
        this.nextToken = nextToken;
        this.leftPrecedence = leftPrecedence;
        assert (functor != null);
        assert (!hasLeft || leftPrecedence != null);
    }

    public void endPreParse(ParseState state, Token nextToken, Integer leftPrecedence) {
        this.state = state;
        this.nextToken = nextToken;
        this.leftPrecedence = leftPrecedence;
        assert (!hasLeft || leftPrecedence != null);
    }

    public void endRepetition(Set<RepetitionPattern> endRepetitions, Token nextToken) {
        this.endRepetitions = endRepetitions;
        this.nextToken = nextToken;
    }

    public void startRepetition() {
        this.endRepetitions = Set.of();
    }

    public List<AstElement> elements() {
        return elements.toImmutable();
    }

    public boolean isEmpty() {
        int size = elements.size();
        return (hasLeft ? size - 1 : size) == 0;
    }

    public boolean hasLeft() {
        return hasLeft;
    }

    public void left(AstElement element) {
        elements.add(element);
        hasLeft = true;
    }

    public void add(AstElement element) {
        elements.add(element);
    }

    @SuppressWarnings({"rawtypes", "unchecked"})
    public Node postParse(ParseContext ctx) throws ParseException {
        ParseState next = state;
        if (next != null) {
            state = null;
            next.parse(nextToken, this, Map.of(), false);
        }
        if (functor != null) {
            for (Pair<Token, Token> split : splitted) {
                split.a().connect(split.b());
            }
            for (Pair<Token, Token> merge : merged) {
                merge.a().merge(merge.b());
            }
            List<AstElement> elements = elements();
            Map<Variable, Type> ta = typeArgs.toImmutable();
            Object[] args = functor.args(elements, ta);
            if (!ta.isEmpty()) {
                functor = functor.setBinding((Map) ta);
            }
            Node node = functor.construct(elements, args, this);
            if (hasLeft && args.length == 1 && args[0] instanceof Node arg) {
                if (node.functor().equals(arg.functor())) {
                    addException(new ParseException("Circular object construction, caused by " + functor, elements));
                    return null;
                }
            }
            if (Type.ROOT.isAssignableFrom(node.type())) {
                node.init(parser.knowledgeBase());
            }
            return node;
        }
        return null;
    }

    @Override
    public String toString() {
        return elements().toString().substring(4);
    }

    @Override
    public void addException(ParseException exception) throws ParseException {
        parser.addException(exception);
    }

    @Override
    public List<ParseException> exceptions() {
        return parser.exceptions();
    }

    public void putTypeArg(Variable arg, Type val) {
        typeArgs.put(arg, val);
    }

    public Type getTypeArg(Variable arg) {
        return typeArgs.get(arg);
    }

}

File: web-ts/src/syntax/PatternResult.ts
/**
 * PatternResult - intermediate pattern match results.
 * @JAVA_REF org.modelingvalue.nelumbo.syntax.PatternResult
 */

import { List, Map, Set } from 'immutable';
import type { Token } from './Token';
import type { AstElement } from '../AstElement';
import { Type } from '../Type';
import { Variable } from '../Variable';
import { Node } from '../Node';
import { Functor } from '../patterns/Functor';
import { RepetitionPattern } from '../patterns/RepetitionPattern';
import type { ParseContext } from './ParseContext';
import { ParseState } from './ParseState';
import { ParseException } from './ParseException';
import type { Parser } from './Parser';
import type { ParseExceptionHandler } from '../KnowledgeBase';
import { Transform } from '../Transform';

/**
 * PatternResult - holds the intermediate result of pattern matching.
 * @JAVA_REF org.modelingvalue.nelumbo.syntax.PatternResult
 */
export class PatternResult implements ParseExceptionHandler {
  private _elements: AstElement[] = [];
  private readonly _parser: Parser;
  private readonly _context: ParseContext;
  private _splitted: Array<[Token, Token]> = [];
  private _merged: Array<[Token, Token]> = [];
  private _typeArgs: Map<Variable, Type> = Map();

  private _functor: Functor | null = null;
  private _state: ParseState | null = null;
  private _leftPrecedence: number | null = null;
  private _endRepetitions: Set<RepetitionPattern> = Set();
  private _nextToken: Token | null = null;
  private _hasLeft: boolean = false;

  constructor(parser: Parser, context: ParseContext) {
    this._parser = parser;
    this._context = context;
  }

  // @JAVA_REF PatternResult.addSplit(Token, Token)
  addSplit(original: Token, split: Token): Token {
    this._splitted.push([original, split]);
    return split;
  }

  // @JAVA_REF PatternResult.addMerge(Token, Token)
  addMerge(original: Token, merge: Token): Token {
    this._merged.push([original, merge]);
    return merge;
  }

  // @JAVA_REF PatternResult.parser()
  parser(): Parser {
    return this._parser;
  }

  // @JAVA_REF PatternResult.context()
  context(): ParseContext {
    return this._context;
  }

  // @JAVA_REF PatternResult.functor()
  functor(): Functor | null {
    return this._functor;
  }

  // @JAVA_REF PatternResult.state()
  state(): ParseState | null {
    return this._state;
  }

  // @JAVA_REF PatternResult.endRepetitions()
  endRepetitions(): Set<RepetitionPattern> {
    return this._endRepetitions;
  }

  // @JAVA_REF PatternResult.nextToken()
  nextToken(): Token | null {
    return this._nextToken;
  }

  // @JAVA_REF PatternResult.leftPrecedence()
  leftPrecedence(): number | null {
    return this._leftPrecedence;
  }

  // @JAVA_REF PatternResult.endPostParse(Functor, Token, Integer)
  endPostParse(functor: Functor, nextToken: Token | null, leftPrecedence: number | null): void {
    this._endRepetitions = Set();
    this._functor = functor;
    this._nextToken = nextToken;
    this._leftPrecedence = leftPrecedence;
    // assert (functor != null)
    // assert (!hasLeft || leftPrecedence != null)
  }

  // @JAVA_REF PatternResult.endPreParse(ParseState, Token, Integer)
  endPreParse(state: ParseState, nextToken: Token | null, leftPrecedence: number | null): void {
    this._state = state;
    this._nextToken = nextToken;
    this._leftPrecedence = leftPrecedence;
    // assert (!hasLeft || leftPrecedence != null)
  }

  // @JAVA_REF PatternResult.endRepetition(Set, Token)
  endRepetition(endRepetitions: Set<RepetitionPattern>, nextToken: Token | null): void {
    this._endRepetitions = endRepetitions;
    this._nextToken = nextToken;
  }

  // @JAVA_REF PatternResult.startRepetition()
  startRepetition(): void {
    this._endRepetitions = Set();
  }

  // @JAVA_REF PatternResult.elements()
  elements(): List<AstElement> {
    return List(this._elements);
  }

  // @JAVA_REF PatternResult.hasLeft()
  hasLeft(): boolean {
    return this._hasLeft;
  }

  // @JAVA_REF PatternResult.left(AstElement)
  left(element: AstElement): void {
    this._elements.push(element);
    this._hasLeft = true;
  }

  // @JAVA_REF PatternResult.add(AstElement)
  add(element: AstElement): void {
    this._elements.push(element);
  }

  // @JAVA_REF PatternResult.removeLast()
  removeLast(): void {
    this._elements.pop();
  }

  // @JAVA_REF PatternResult.nrOfExceptions() (from ParseExceptionHandler)
  nrOfExceptions(): number {
    return this.exceptions().size;
  }

  // @JAVA_REF PatternResult.postParse(ParseContext)
  postParse(_ctx: ParseContext): Node | null {
    const next = this._state;
    if (next !== null) {
      this._state = null;
      next.parse(this._nextToken, this, Map(), false);
    }
    if (this._functor !== null) {
      for (const [original, split] of this._splitted) {
        original.connect(split);
      }
      for (const [original, merge] of this._merged) {
        original.merge(merge);
      }
      const elements = this.elements();
      const ta = this._typeArgs;
      const args = this._functor.functorArgs(elements, ta);
      let functor = this._functor;
      if (!ta.isEmpty()) {
        functor = functor.setBinding(ta as Map<Variable, unknown>) as Functor;
      }
      const node = functor.construct(elements, args);
      // Circular reference check
      if (this._hasLeft && args.length === 1 && args[0] instanceof Node) {
        const arg = args[0] as Node;
        if (node.functor() !== null && node.functor()!.equals(arg.functor())) {
          this.addException(ParseException.fromElements('Circular object construction, caused by ' + functor, node));
          return null;
        }
      }
      if (Type.ROOT.isAssignableFrom(node.type())) {
        const kb = this._parser.knowledgeBase();
        // Java virtual dispatch: node.init(kb)
        // Transform.init → addTransform only (no transform application)
        // Functor.init → register only (already done via createFunctor)
        // Other subclasses have their own init overrides
        // Only base Node.init applies transforms
        if (node instanceof Transform) {
          node.initInKb(kb);
        } else if (node instanceof Functor) {
          // Functor.init in Java calls register() — already done via createFunctor
          // Don't apply transforms to Functor nodes
        } else {
          // Base Node.init(kb) - apply transforms
          for (const transform of kb.getTransforms(node)) {
            transform.rewrite(transform.source(), node, kb);
          }
        }
      }
      return node;
    }
    return null;
  }

  // @JAVA_REF PatternResult.toString()
  toString(): string {
    return this.elements().toArray().toString();
  }

  // @JAVA_REF PatternResult.addException(ParseException)
  addException(exception: ParseException): void {
    this._parser.addException(exception);
  }

  // @JAVA_REF PatternResult.exceptions()
  exceptions(): List<ParseException> {
    return this._parser.exceptions();
  }

  // @JAVA_REF PatternResult.putTypeArg(Variable, Type)
  putTypeArg(arg: Variable, val: Type): void {
    this._typeArgs = this._typeArgs.set(arg, val);
  }

  // @JAVA_REF PatternResult.getTypeArg(Variable)
  getTypeArg(arg: Variable): Type | undefined {
    return this._typeArgs.get(arg);
  }
}
